server/src/main/java/org/opensearch/plugins/Plugin.java

```diff
@@ -278,4 +279,10 @@ public Collection<IndexSettingProvider> getAdditionalIndexSettingProviders() {
     public Optional<SecureSettingsFactory> getSecureSettingFactory(Settings settings) {
         return Optional.empty();
     }
+
+    /**
+     * Called to set the {@link TelemetryService} and plugins can use it for generating the instrumentations.
+     * @param telemetryService telemetry service
+     */
+    public void setTelemetryService(TelemetryService telemetryService) {}
```

Comment:
The telemetry comes from the plugin so we are now introducing the dependency between all and every plugin on telemetry plugin, this is not the way to approach the problem.

```suggestion
@ExperimentalApi
public interface TelemetryAwarePlugin {
    default Optional<Tracer> getTracer() {
        return Optional.empty();
    }    
    
    default Optional<MetricsRegistry> getMetricsRegistry() {
        return Optional.empty();
    }
}
```

The services should not be pushed to the plugin(s) but provided on pull based model. The purpose of `TelemetryService` is unclear - why do we need it instead of giving access to `Tracer` or/and `MetricsRegistry` as we do in all other places?

Reply:
@reta, thanks for reviewing this.

>> The telemetry comes from the plugin so we are now introducing the dependency between all and every plugin on telemetry plugin, this is not the way to approach the problem.

I agree and didn't find a straightforward way to accomplish that. My thinking was that Telemetry is a fundamental concept that includes a default Noop implementation. OpenTelemetry (Otel) provides its non-default concrete implementation through plugins, but telemetry as a concept overall belongs to Opensearch core and can be injected into all plugins through the PluginService.

I've observed other plugins passing objects to plugins using existing services like ClusterService via the Plugin::createComponent method, which I didn't find ideal or even intended for that purpose. I'm looking for the best alternative to solve this issue.


>> The purpose of TelemetryService is unclear - why do we need it instead of giving access to Tracer or/and MetricsRegistry as we do in all other places?

Earlier, there was a request highlighting that passing both the tracer and metrics registry separately is cumbersome and requires a single object. I have introduced a solution here and will later replace the existing code as well, moving it to the TelemetryService. In future, let's say we introduce logging as well then changes at multiple places is not needed.

Reply:
I agree and didn't find a straightforward way to accomplish that. My thinking was that Telemetry is a fundamental concept that includes a default Noop implementation. OpenTelemetry (Otel) provides its non-default concrete implementation through plugins, but telemetry as a concept overall belongs to Opensearch core and can be injected into all plugins through the PluginService.

I thought of this approach (and was very much aligned) but didn't understand how will I pass the implementation of tracer and metricRegistry from core to the implementing plugins. I think instead of getter, setter should help here and should be called only for TelemetryAwarePlugins.

>> The purpose of TelemetryService is unclear - why do we need it instead of giving access to Tracer or/and MetricsRegistry as we do in all other places?

Earlier, there was a request highlighting that passing both the tracer and metrics registry separately is cumbersome and requires a single object. I have introduced a solution here and will later replace the existing code as well, moving it to the TelemetryService. In future, let's say we introduce logging as well then changes at multiple places is not needed.

Reply:
> I agree and didn't find a straightforward way to accomplish that

With separate plugin interface, it should be feasible: the OpenSearch core may need to account for initialization sequence and prioritize the telemetry plugin (which should be prohibited from implementing  `TelemetryAwarePlugin`)

Reply:
Yes, we can do that. But it would look something like this.

```
@ExperimentalApi
public interface TelemetryAwarePlugin {
    default void setTracer(Tracer tracer) {
     }    
    
    default void setMetricsRegistry() {
       
    }
}
```

Node.java

```

pluginsService.filterPlugins(TelemetryAwarePlugin.class).stream().forEach(p -> { 
p.setTracer(tracer);
p.setMetricsRegistry(metricsRegistry);
});

```

Reply:
I disagree, the setters is difficult to reason about: 
- it is difficult to communicate when the setter is called, there is always uncertainty because the value in question could be used **before** it is set
- the value of the setter has to be stored somewhere, we are forcing every single plugin in question to repeat that

With provider model, the provided value is always available.

Reply:
Provider will be definitely helpful and even I don't like the idea of setter and like you mentioned it may create mess. To some extent TelemetryService was helping here. But How to make the provider available to the plugin? Could you please share some example/reference from the code?

Reply:
> Could you please share some example/reference from the code?

Sure, I will try to find the time this week

Reply:
Thanks @reta. I have looked into the plugin code, there are provider implementations where we are providing the values out of plugin to core but not the other way around.

Reply:
@Gaganjuneja apologies for delay, took me longer to get to the issue

>Thanks @reta. I have looked into the plugin code, there are provider implementations where we are providing the values out of plugin to core but not the other way around.

Yes, you are right, we sadly do not support injection of the services to the plugins (I think it could be useful in general but not in scope of this change. So your establish approach to create the dependent components is to use  `Plugin::createComponents` method. 


I would suggest introducing `TelemetryAwarePlugin`, as before, but with different exposed API, basically  `Plugin::createComponents`  + Tracer + MetricsRegistry. 

```
@ExperimentalApi 
public interface TelemetryAwarePlugin {
    public Collection<Object> createComponents(
        Client client,
        ClusterService clusterService,
        ThreadPool threadPool,
        ResourceWatcherService resourceWatcherService,
        ScriptService scriptService,
        NamedXContentRegistry xContentRegistry,
        Environment environment,
        NodeEnvironment nodeEnvironment,
        NamedWriteableRegistry namedWriteableRegistry,
        IndexNameExpressionResolver indexNameExpressionResolver,
        Supplier<RepositoriesService> repositoriesServiceSupplier,
        Tracer getTracer,
        MetricsRegistry metricsRegistry
    ) {
        return Collections.emptyList();
    }
}
```

The `Node` initializer should prioritize `TelemetryPlugin` (so tracer and metric registry become available) and call `TelemetryAwarePlugin::createComponents` for all plugins that do implement this interface.

Reply:
Thanks @reta. That's the only option we have. Here, createComponents still means create the component inside the plugin and expose to the core as per the javadoc comment `Returns components added by this plugin.` and usage. Instead shouldn't we have  something like `TelemetryAwarePlugin::registerTelemetry(Tracer tracer, MetricsRegistry metricsRegistry)`

Reply:
> returns components added by this plugin.

This is fine I think, the list of components could be empty, this is the only injection point we have: it has well defined lifecycle and plugin developers could be assured the needed services are initialized.

Reply:
Ok. Now the second point should we keep tracer and metricsRegistry separate objects or encapsulated inside a TelemetryService. TelemetryService reduces the boilerplate code and avoids changing the constructor multiple times.

Reply:
> TelemetryService reduces the boilerplate code and avoids changing the constructor multiple times.

The TelemetryService is not needed in my opinion, it is man-in-middle abstraction with no value - providing direct access to Tracer and MetricsRegistry is what matters.